### PR TITLE
ENT-10100 Changes required to support recovery of IN_FLIGHT transactions.

### PR DESCRIPTION
**Pull request by colljos**
Additional support for Enterprise Ledger Recovery:
- addition of new flow: `FetchRecoverableTransactionsFlow`
- expose new public function `getTransactionWithStatus` (and remove equivalent internal `getTransactionInternal`)

### CODE REVIEW

File: node/src/main/kotlin/net/corda/node/services/DbTransactionsResolver.kt [hunk 1]
```diff
@@ -56,7 +57,7 @@ class DbTransactionsResolver(private val flow: ResolveTransactionsFlow) : Transa
             }
 
             // Request the standalone transaction data (which may refer to things we don't yet have).
-            val (existingTxIds, downloadedTxs) = fetchRequiredTransactions(Collections.singleton(nextRequests.first())) // Fetch first item only
+            val (existingTxIds, downloadedTxs) = fetchRequiredTransactions(Collections.singleton(nextRequests.first()), recoveryMode) // Fetch first item only
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/services/DbTransactionsResolver.kt]: Maybe you can just ignore the value of recoveryMode in most of this class and just throw an error if `downloadDependencies` is called with it true.

File: core/src/main/kotlin/net/corda/core/internal/FetchDataFlow.kt [hunk 2]
```diff
@@ -267,8 +267,8 @@ class FetchAttachmentsFlow(requests: Set<SecureHash>,
  * Authorisation is accorded only on valid ancestors of the root transaction.
  * Note that returned transactions are not inserted into the database, because it's up to the caller to actually verify the transactions are valid.
  */
-class FetchTransactionsFlow(requests: Set<SecureHash>, otherSide: FlowSession) :
-        FetchDataFlow<SignedTransaction, SignedTransaction>(requests, otherSide, DataType.TRANSACTION) {
+class FetchTransactionsFlow(requests: Set<SecureHash>, otherSide: FlowSession, dataType: DataType = DataType.TRANSACTION) :
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/internal/FetchDataFlow.kt]: Think you might need to maintain constructors so perhaps add JVM overloads

File: core/src/main/kotlin/net/corda/core/internal/FetchDataFlow.kt [hunk 3]
```diff
@@ -82,7 +82,7 @@ sealed class FetchDataFlow<T : NamedByHash, in W : Any>(
     )
     @CordaSerializable
     enum class DataType {
-        TRANSACTION, ATTACHMENT, PARAMETERS, BATCH_TRANSACTION, UNKNOWN
+        TRANSACTION, TRANSACTION_RECOVERY, ATTACHMENT, PARAMETERS, BATCH_TRANSACTION, UNKNOWN
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/internal/FetchDataFlow.kt]: Do NOT put things in the middle of an enum.  Put it at the end.  Has jeopardy for anything using ordinal etc.  Maybe serialisation etc.

Comment (colljos) [reply core/src/main/kotlin/net/corda/core/internal/FetchDataFlow.kt]: Good point! It just looked so nice next to its silbling ;-)

File: node/src/main/kotlin/net/corda/node/services/DbTransactionsResolver.kt [hunk 4]
```diff
@@ -99,10 +100,9 @@ class DbTransactionsResolver(private val flow: ResolveTransactionsFlow) : Transa
...
         val sortedDependencies = checkNotNull(this.sortedDependencies)
         logger.trace { "Recording ${sortedDependencies.size} dependencies for ${flow.txHashes.size} transactions" }
-        val transactionStorage = flow.serviceHub.validatedTransactions as WritableTransactionStorage
         for (txId in sortedDependencies) {
             // Retrieve and delete the transaction from the unverified store.
-            val (tx, txStatus) = checkNotNull(transactionStorage.getTransactionInternal(txId)) {
+            val (tx, txStatus) = checkNotNull(flow.serviceHub.validatedTransactions.getTransactionWithStatus(txId)) {
                 "Somehow the unverified transaction ($txId) that we stored previously is no longer there."
             }
             if (txStatus != TransactionStatus.VERIFIED) {
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/services/DbTransactionsResolver.kt]: If it is IN_FLIGHT, you can't use `recordTransactions()`, surely?   But is the expectation that you will never get an IN_FLIGHT in this implementation (I think that is true, BTW) and thus okay?

File: core/src/main/kotlin/net/corda/core/internal/FetchDataFlow.kt [hunk 5]
```diff
@@ -267,12 +268,28 @@ class FetchAttachmentsFlow(requests: Set<SecureHash>,
...
 
+// Used by Enterprise Ledger Recovery
+class FetchRecoverableTransactionsFlow @JvmOverloads constructor(requests: Set<SecureHash>, otherSide: FlowSession, dataType: DataType = DataType.TRANSACTION_RECOVERY) :
+        FetchDataFlow<SignedTransactionWithStatus, SignedTransactionWithStatus>(requests, otherSide, dataType) {
+
+    override fun load(txid: SecureHash): SignedTransactionWithStatus? = serviceHub.validatedTransactions.getTransactionWithStatus(txid)
+}
+
+@CordaSerializable
+data class SignedTransactionWithStatus(
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/internal/FetchDataFlow.kt]: Hmmm.... I don't think we should have an internal type returned from a non-internal function (i.e. `getTransactionWithStatus()` ).  Move the to the package where that is defined etc?

File: core/src/main/kotlin/net/corda/core/node/services/TransactionStorage.kt [hunk 6]
```diff
@@ -17,6 +18,11 @@ interface TransactionStorage {
      */
     fun getTransaction(id: SecureHash): SignedTransaction?
 
+    /**
+     * Return the transaction with its status for the given [id], or null if no such transaction exists.
+     */
+    fun getTransactionWithStatus(id: SecureHash): SignedTransactionWithStatus?
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/node/services/TransactionStorage.kt:26]: See comment elsewhere about returning internal type.

File: node/src/main/kotlin/net/corda/node/services/DbTransactionsResolver.kt (hunk 4 above)
Comment (rick-r3) [reply node/src/main/kotlin/net/corda/node/services/DbTransactionsResolver.kt]: Perhaps we should be more explicit and throw if you get an IN_FLIGHT?

Comment (colljos) [reply node/src/main/kotlin/net/corda/node/services/DbTransactionsResolver.kt]: We can't throw an exception for `IN_FLIGHT` as we actually recover `IN_FLIGHT` transactions in Ledger Recovery mode.
As per a previous comment you made (in the ENT PR), I think we should do nothing.

Comment (colljos) [reply node/src/main/kotlin/net/corda/node/services/DbTransactionsResolver.kt]: I've changed the status check to:
`(txStatus == TransactionStatus.UNVERIFIED)`

**Review by rick-r3 — APPROVED** [review]

**Pull request merged**